### PR TITLE
Introduce Ruby27 pattern matching limited support

### PR DESCRIPTION
**Pull request by Morriar**
### Motivation

Ruby 27 [introduces pattern matching](https://www.ruby-lang.org/en/news/2019/12/25/ruby-2-7-0-released/), this PR is a first step to support it.

The main goal is to be able to parse and analyze Ruby code without crashing.

### What's included?

* Sorbet can parse pattern matching expressions
* Sorbet can desugar them
* Sorbet can type the content of each `in` of a `case`
* Sorbet can see the variables declared inside a pattern
* But... the variables from the pattern are untyped at this point.

### How does it work?

Once the pattern matching expression is parsed we desugar it as a if plus some magic.

Be the example from the Ruby news article linked above:

```rb
# typed: true

json = "..."

case JSON.parse(json, symbolize_names: true)
in {name: "Alice", children: [{name: "Bob", age: age}]}
  T.reveal_type(age) # T.untyped
else
  puts "No match"
end
```

We desugar it as:

```rb
class <emptyTree><<C <root>>> < (::<todo sym>)
  json = "..."

  begin
    <assignTemp>$1 = <emptyTree>::<C JSON>.parse(json, {:"symbolize_names" => true})
    if ::<Magic>.<pattern-match>(<assignTemp>$1)
      begin
        age = ::<Magic>.<pattern-match-var>()
        <emptyTree>::<C T>.reveal_type(age)
      end
    else
      <self>.puts("No match")
    end
  end
end
```

And Sorbet says (note that the matched var is correctly defined but untyped):

```
test.rb:7: Revealed type: T.untyped https://srb.help/7014
     7 |  T.reveal_type(age) # T.untyped
          ^^^^^^^^^^^^^^^^^^
  From:
    test.rb:6:
     6 |in {name: "Alice", children: [{name: "Bob", age: age}]}
                                                         ^^^
Errors: 1
```

### What's missing?

* The pattern itself is not desugared at this point, I didn't want to create the nodes and everything until we know what we want to do with them. Once desugared the pattern should be passed to ` ::<Magic>.<pattern-match>(<assignTemp>$1, PATTERN)`. We can also add a magic for each kind of pattern (`<pattern-match-const>`, `<pattern-match-array>`, `<pattern-match-hash>`).

* The variables from the pattern are initialized with `var = ::<Magic>.<pattern-match-var>()` so we loose the type information (hence the `untyped`). We could do some magic to initialize the array from the `case` expression like: `var = expr[var_index or var_key]` but this involves handling rest `args`, `kwargs`, `kwnil`... So a lot of work and this PR is already big.

* Same for the `as` variable in `in Foo => x`. It is initialized as `var = ::<Magic>.<pattern-match-as>()`. Here the type should come from the pattern itself and since the pattern can contain types we can also use them.

* The guards is not desugared either. We could pass it to ` ::<Magic>.<pattern-match>(<assignTemp>$1, PATTERN, GUARD)` or introduce a new `if` in the AST.

### So does it work?

Yes!

We can merge today and enhance later.

### Test plan

There is a lot of syntax included, see included automated tests for more details.

### CODE REVIEW

**Comment by Morriar** [comment]
Rebased on master

File: ast/desugar/Desugar.cc [hunk 1]
```diff
@@ -404,6 +404,41 @@ TreePtr desugarMlhs(DesugarContext dctx, core::LocOffsets loc, parser::Mlhs *lhs
     return MK::InsSeq(loc, std::move(stats), MK::Local(loc, tempRhs));
 }
 
+// Map all MatchVars used in `pattern` to local variables initialized from magic calls
+void desugarPatternMatchingVars(InsSeq::STATS_store &vars, DesugarContext dctx, unique_ptr<parser::Node> &node) {
+    if (auto var = parser::cast_node<parser::MatchVar>(node.get())) {
+        auto loc = var->loc;
+        auto recv = MK::Constant(loc, core::Symbols::Magic());
+        auto val = MK::Send0(loc, std::move(recv), core::Names::patternMatchVar());
```
Comment (jez) [inline ast/desugar/Desugar.cc:412]: This intrinsic just returns `T.untyped`, right? Is there any point to it? It takes no arguments so it doesn't even keep track of enough original state for a later part of the pipeline to be able to learn the type.

With that in mind, can we make this more obvious that it's doing something wrong? Like `ast::MK::Unimplemented(loc)`? That helper makes an untyped value just the same.

File: ast/desugar/Desugar.cc [hunk 2]
```diff
@@ -404,6 +404,41 @@ TreePtr desugarMlhs(DesugarContext dctx, core::LocOffsets loc, parser::Mlhs *lhs
...
+        vars.emplace_back(MK::Assign(loc, var->name, std::move(val)));
+    } else if (auto rest = parser::cast_node<parser::MatchRest>(node.get())) {
+        desugarPatternMatchingVars(vars, dctx, rest->var);
+    } else if (auto pair = parser::cast_node<parser::Pair>(node.get())) {
+        desugarPatternMatchingVars(vars, dctx, pair->value);
+    } else if (auto as_pattern = parser::cast_node<parser::MatchAs>(node.get())) {
+        auto name = parser::cast_node<parser::MatchVar>(as_pattern->as.get())->name;
+        auto recv = MK::Constant(as_pattern->as->loc, core::Symbols::Magic());
+        auto val = MK::Send0(as_pattern->as->loc, std::move(recv), core::Names::patternMatchAs());
```
Comment (jez) [inline ast/desugar/Desugar.cc:421]: What is the difference between `<Magic>.<pattern-match-var>` and `<Magic>.<pattern-match-as>`?

Could we make them both use `ast::MK::Unimplemented`?

File: ast/desugar/Desugar.cc [hunk 3]
```diff
@@ -1885,6 +1920,44 @@ TreePtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what) {
...
+                    auto body = node2TreeImpl(dctx, std::move(inPattern->body));
+
+                    // Desugar match variables found inside the pattern
+                    InsSeq::STATS_store vars;
+                    desugarPatternMatchingVars(vars, dctx, pattern);
+                    if (!vars.empty()) {
+                        body = MK::InsSeq(pattern->loc, std::move(vars), std::move(body));
+                    }
+
+                    // Call `Magic.<pattern-match>(expr, TODO pattern, TODO guard)`
```
Comment (jez) [inline ast/desugar/Desugar.cc:1949]: This comment mentions three arguments, but I only see one argument passed to `<Magic>.<pattern-match>`?

Comment (jez) [reply ast/desugar/Desugar.cc:1949]: Also ditto above—can we use `ast::MK::Unimplemented` for the pattern match expr if we're not going to implement it with an intrinsic or something else?

File: parser/Builder.cc [hunk 4]
```diff
@@ -1271,6 +1448,52 @@ class Builder::Impl {
...
+        }
+
+        driver_->pattern_hash_keys.declare(name);
+    }
+
+    void checkLVarName(std::string name, core::LocOffsets loc) {
+        std::regex lvar_regex("^[a-z_][a-zA-Z0-9_]*$");
+        if (!std::regex_match(name, lvar_regex)) {
+            error(ruby_parser::dclass::PatternLVarName, loc, name);
+        }
```
Comment (jez) [inline parser/Builder.cc:1476]: ooc, why is this not handled by the grammar?

File: parser/Builder.cc [hunk 5]
```diff
@@ -558,6 +596,11 @@ class Builder::Impl {
         return make_unique<Const>(tokLoc(name), nullptr, gs_.enterNameConstant(name->string()));
     }
 
+    unique_ptr<Node> const_pattern(unique_ptr<Node> const_, const token *begin, unique_ptr<Node> pattern,
```
Comment (jez) [inline parser/Builder.cc:599]: I believe most of the methods in this file are `camelCase`. Would you mind following that convention?

**Review by jez — COMMENTED** [review]
I left some comments. This change looks mostly fine with me. I'd prefer to be more explicit about the fact that everything is unimplemented with `ast::MK::Unimplemented`.

We will have to test this on Stripe's codebase, but tother than that we can probably merge this in.

Thanks!

**Comment by jez** [comment]
http://go/builds/bui_IUfOVLYLuvNmsE

File: ast/desugar/Desugar.cc (hunk 1 above)
Comment (Morriar) [reply ast/desugar/Desugar.cc]: Replaced by `RaiseUnimplemented`.

File: ast/desugar/Desugar.cc (hunk 2 above)
Comment (Morriar) [reply ast/desugar/Desugar.cc]: `match-var` was meant to access a matched variable like `a`, while `match-as` was to access a bound variable like `a => Integer`.

Replaced both by `RaiseUnimplemented`.

File: ast/desugar/Desugar.cc (hunk 3 above)
Comment (Morriar) [reply ast/desugar/Desugar.cc]: Replaced by `RaiseUnimplemented`.

File: parser/Builder.cc (hunk 4 above)
Comment (Morriar) [reply parser/Builder.cc:1476]: Not sure, I would say to avoid conflicts?

File: parser/Builder.cc (hunk 5 above)
Comment (Morriar) [reply parser/Builder.cc:599]: Before my changes:

38 snake_case: ``` array_pattern,
    call_method,
    case_match,
    const_pattern,
    def_class,
    def_sclass,
    hash_pattern,
    if_guard,
    in_match,
    in_pattern,
    line_literal,
    loop_while,
    loop_while_mod,
    multi_assign,
    multi_lhs,
    multi_lhs1,
    not_op,
    nth_ref,
    op_assign,
    p_ident,
    pair_keyword,
    pair_quoted,
    range_exclusive,
    range_inclusive,
    rational_complex,
    regexp_compose,
    regexp_options,
    rescue_body,
    splat_mlhs,
    string_compose,
    string_internal,
    symbol_compose,
    symbol_internal,
    symbols_compose,
    unary_op,
    unless_guard,
    words_compose,
    xstring_compose ```

28 CamelCase: ```attrAsgn,
    backRef,
    beginBody,
    beginKeyword,
    binaryOp,
    blockPass,
    callLambda,
    conditionMod,
    constFetch,
    constGlobal,
    constOpAssignable,
    dedentString,
    fileLiteral,
    floatComplex,
    indexAsgn,
    keywordBreak,
    keywordDefined,
    keywordNext,
    keywordRedo,
    keywordRetry,
    keywordReturn,
    keywordSuper,
    keywordYield,
    keywordZsuper,
    logicalAnd,
    logicalOr,
    loopUntil,
    loopUntil_mod```

I went with the most used.

Comment (jez) [reply parser/Builder.cc:599]: makes sense, thanks!

**Review by jez — APPROVED** [review]
Thanks!

**Pull request merged**